internal/pkg/template/repository/manager/repository.go

```diff
@@ -0,0 +1,231 @@
+package manager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/filesystem"
+	"github.com/keboola/keboola-as-code/internal/pkg/git"
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/template"
+	"github.com/keboola/keboola-as-code/internal/pkg/template/repository"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils"
+	pullOp "github.com/keboola/keboola-as-code/pkg/lib/operation/repository/pull"
+	loadTemplateOp "github.com/keboola/keboola-as-code/pkg/lib/operation/template/load"
+	loadRepositoryOp "github.com/keboola/keboola-as-code/pkg/lib/operation/template/repository/load"
+)
+
+// CachedRepository combines low-level git.Repository and high-level *repository.Repository.
+// It adds caching support and coordinates template loading.
+type CachedRepository struct {
+	d dependencies
+
+	git  git.Repository // *git.LocalRepository (type=dir) or *git.RemoteRepository (type=git)
+	repo *repository.Repository
+
+	templates     map[string]*template.Template
+	templatesInit *singleflight.Group // each template is load only once
+	templatesLock *sync.RWMutex       // provides atomic access to the templates field
+
+	unlockFn git.RepositoryFsUnlockFn // unlocks underlying FS, called on free()
+	freeLock *sync.RWMutex            // prevents cleanup of the repository while it is in use, see markInUse and free methods
+}
+
+// UnlockFn callback is returned by Manager. It must be called when the Cached Repository is no longer in use.
+type UnlockFn func()
+
+// newCachedRepository inits CachedRepository and preload all templates.
+func newCachedRepository(ctx context.Context, d dependencies, gitRepo git.Repository, unlockFn git.RepositoryFsUnlockFn, tmplRepo *repository.Repository) *CachedRepository {
+	r := &CachedRepository{
+		d:             d,
+		git:           gitRepo,
+		repo:          tmplRepo,
+		templates:     make(map[string]*template.Template),
+		templatesInit: &singleflight.Group{},
+		templatesLock: &sync.RWMutex{},
+		unlockFn:      unlockFn,
+		freeLock:      &sync.RWMutex{},
+	}
+
+	// Reload all templates, error in a template is logged
+	_ = r.loadAllTemplates(ctx)
+	return r
```

Comment:
Zamerne sa tu ignoruje chyba, ak bude jedna template nevalidna, tak ostatne sa mozu pouzivat dalej, do logov nam pride chyba. Ak sa to bude diat casto, tak logy presmerujeme na autora sablony :) 

Je to lepsie ako teraz, ked sme nevedeli o chybe, kym nevyskocila uzivatelovi. 
Toto zachyti vsetky chyby az po aplikovanie sablony (to uz nie).